test/framework/clustervalidator.go

```diff
@@ -0,0 +1,92 @@
+package framework
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/retrier"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type clusterValidation = func(client client.Client, config *cluster.Config) error
+
+type validationOpt struct {
+	timeout  time.Duration
+	validate clusterValidation
+}
+
+// ClusterValidator is responsible for checking if a cluster is valid against the cluster config that is provided.
+type ClusterValidator struct {
+	clusterConfig *cluster.Config
+	client        client.Client
+	validations   []validationOpt
+}
+
+func (c *ClusterValidator) withValidation(validate clusterValidation, timeout time.Duration) {
+	c.validations = append(c.validations, validationOpt{
+		timeout:  timeout,
+		validate: validate,
+	})
+}
+
+// ValidateCluster runs through the set clusterValidations returns an error if any of them fail after a number of retries.
+func (c *ClusterValidator) ValidateCluster() error {
+	for _, validation := range c.validations {
+		err := retrier.New(validation.timeout).Retry(func() error {
+			return validation.validate(c.client, c.clusterConfig)
+		})
+
+		if err != nil {
+			return fmt.Errorf("validation faild %v", err)
+		}
+
+	}
+
+	return nil
+}
+
+// NewDefaultClusterValidator returns a cluster validator pre-loaded with validations by default.
+func NewDefaultClusterValidator(client client.Client, config *cluster.Config) ClusterValidator {
+	cv := ClusterValidator{
+		clusterConfig: config,
+		client:        client,
+	}
+
+	cv.withValidation(validateObjects, 5*time.Minute)
+	cv.withValidation(validateControlPlaneNodes, 5*time.Minute)
```

Comment:
what's the purpose of having a system to dynamically register validations if we hardcode the validations in the constructor?

I would vote to decouple both, having a constructor that just returns the validator and allowing to add validations later. And if you want, we can have a wrapper that init and register the default validations.

Reply:
My thought was that we could have a constructor that returns a validator without validations as well, in addition to one that's already pre-loaded. I can see the confusion though, I think having a wrapper that inits and registers defaults validations is where I'll take this.